pkg/hub/taints/controller.go

```diff
@@ -0,0 +1,166 @@
+package taints
+
+import (
+	"context"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	clientset "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	informerv1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	listerv1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	v1 "open-cluster-management.io/api/cluster/v1"
+	"time"
+)
+
+const (
+	NoCondition      = "NoManagedClusterConditionAvailable"
+	ConditionUnknown = "ManagedClusterConditionAvailableUnknown"
+	ConditionFalse   = "ManagedClusterConditionAvailableFalse"
+)
+
+// taintsController
+type taintsController struct {
+	kubeClient    kubernetes.Interface
+	clusterClient clientset.Interface
+	clusterLister listerv1.ManagedClusterLister
+	eventRecorder events.Recorder
+}
+
+// NewTaintsController creates a new taints controller
+func NewTaintsController(
+	kubeClient kubernetes.Interface,
+	clusterClient clientset.Interface,
+	clusterInformer informerv1.ManagedClusterInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &taintsController{
+		kubeClient:    kubeClient,
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("taints-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("taintsController", recorder)
+}
+
+func (c *taintsController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	managedClusterName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling Taints %s", managedClusterName)
+	managedCluster, err := c.clusterLister.Get(managedClusterName)
+	if errors.IsNotFound(err) {
+		// Spoke cluster not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	managedCluster = managedCluster.DeepCopy()
+	newTaints := managedCluster.Spec.Taints
+
+	switch {
+	case meta.IsStatusConditionTrue(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable):
+		newTaints, _ = c.deleteTaintAndJudgeExist(newTaints, "",
```

Comment:
would you split this method into two methods? one is used to delete and the other is used to judge exist, I think it will be more clear,
e.g. here will be `c.deleteTaints(newTaints , v1.ManagedClusterTaintUnreachable, v1.ManagedClusterTaintUnavailable)`

Reply:
But splitting into two methods will increase the time complexity.